nutils/_util.py

```diff
@@ -778,4 +778,171 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def _reduce(obj):
+    'helper function for deep_replace_property and shallow_replace'
+
+    if obj.__class__ in (tuple, list, dict, set, frozenset):
+        if not obj: # empty containers need not be entered
+            raise Exception('refusing to enter empty container')
+        return lambda *x, T=type(obj): T(x), tuple(obj if not isinstance(obj, dict) else obj.items())
+    return obj.__reduce__()
+
+
+class deep_replace_property:
+    '''decorator for deep object replacement
+
+    Generates a cached property for deep replacement of reduceable objects,
+    based on a callable that is applied depth first and recursively on
+    individual constructor arguments. Intermediate values are stored in the
+    attribute by the same name of any object that is a decendent of the class
+    that owns the property.
+
+    Args
+    ----
+    func
+        Callable which maps an object onto a new object, or `None` if no
+        replacement is made. It must have precisely one positional argument for
+        the object.
+    '''
+
+    def __init__(self, func):
+        self.func = func
+
+    def __set_name__(self, owner, name):
+        self.owner = owner
+        self.name = name
+
+    def __set__(self, obj, value):
+        raise AttributeError("can't set attribute")
+
+    def __delete__(self, obj):
+        raise AttributeError("can't delete attribute")
+
+    def __get__(self, obj, objtype=None):
+        fstack = [obj] # stack of unprocessed objects and command tokens
+        rstack = [] # stack of processed objects
+        recreate = [] # stack of constructor functions
+        originals = [] # stack of original objects to cache new value into
+
+        while fstack:
+            obj = fstack.pop()
+
+            if obj is recreate: # recreate object from rstack
+                f, nargs = recreate.pop()
+                r = f(*[rstack.pop() for _ in range(nargs)])
+                if isinstance(r, self.owner) and (newr := self.func(r)) is not None:
+                    fstack.append(newr) # recursion
+                else:
+                    rstack.append(r)
+
+            elif obj is originals: # store new representation
+                orig = originals.pop()
+                r = rstack[-1]
+                if r is orig: # this may happen if obj is memoizing
+                    r = None # prevent cyclic reference
+                orig.__dict__[self.name] = r
+
+            elif isinstance(obj, self.owner) and self.name in obj.__dict__: # in cache
+                r = obj.__dict__[self.name]
```

Comment:
Suggestion: Combine `self.name in obj.__dict__` and `r = obj.__dict__[self.name]` as follows:

```python
NOT_FOUND = object()
...
elif isinstance(obj, self.owner) and (r := obj.__dict__.get(self.name, NOT_FOUND)) is not NOT_FOUND:
    rstack.append(r if r is not None else obj)
```

or

```python
SELF = object()
...
elif isinstance(obj, self.owner) and (r := obj.__dict__.get(self.name)) is not None:
    rstack.append(obj if r is SELF else r)
```

Reply:
done